### Group Participants
*   Ruzaik:
*   Ravindu:
*   Pulin:
*   Cheyenne Dirckze: COADDS192P-006

### Importing relevant libraries

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import pandas as pd
%matplotlib inline

### Importing datasets

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
path1 = '/content/drive/My Drive/DATA SCIENCE/ML1_GroupAssignment (ABC Telco)/Train data.txt'

In [ ]:
df1 = pd.read_csv(path1)

In [ ]:
path2 = '/content/drive/My Drive/DATA SCIENCE/ML1_GroupAssignment (ABC Telco)/Test data.txt'

In [ ]:
#Carefuly chose header names by comparing against train data with notepad++

df2 = pd.read_csv(path2, header = None, names = ['customerID','gender', 'Senior Citizen', 'tenure', 'InternetService', 'Contract','PaymentMethod','MonthlyCharges', 'TotalCharges'])

### Cleaning up test dataset

In [ ]:
df2.head()

customerID  gender  ...  MonthlyCharges TotalCharges
0  9398-MMQTO    Male  ...           74.45       221.10
1  7619-PLRLP  Female  ...            74.1      3833.95
2  6457-GIRWB    Male  ...           69.35        69.35
3  6508-NJYRO    Male  ...          294.95          NaN
4  1450-SKCVI  Female  ...           73.85      4092.85

[5 rows x 9 columns]

Visible skew of data from tenure column

In [ ]:
df2.isnull().sum()

customerID           0
gender               0
Senior Citizen       0
tenure               0
InternetService      0
Contract             0
PaymentMethod        0
MonthlyCharges       0
TotalCharges       273
dtype: int64

Skew results in 273 missing values for 'Total charges' column

In [ ]:
a= (df2.isnull().sum()['TotalCharges']/df2.shape[0])*100
print('Missing value percentage:', a, '%')

Missing value percentage: 19.941563184806427 %


In [ ]:
#dropping rows containing missing values to save time cleaning data
df2.dropna(inplace = True)

In [ ]:
#Checking for extaordinary unique values in all columns
for i in df2.columns:
    print(df2[i].unique())

['9398-MMQTO' '7619-PLRLP' '6457-GIRWB' ... '1941-HOSAM' '9110-HSGTV'
 '6171-ZTVYB']
['Male' 'Female']
[0 1]
['3' '53' '1' '56' '5' '48' '58' '10' '71' '65' '28 phone service' '67'
 '35' '72 phone service' '61 phone service' '1 phone service' '70'
 '48 phone service' '68' '47' '32 phone service' '13' '12' '9' '38' '42'
 '24' '27' '49 phone service' '61' '50 phone service' '25' '22' '4' '18'
 '51' '62' '16 phone service' '8' '72' '23' '31 phone service'
 '37 phone service' '30' '20' '36' '50' '57' '41' '3 phone service' '64'
 '26 phone service' '40' '34' '63' '44' '6 phone service' '33' '43' '6'
 '59' '14' '19' '2' '57 phone service' '66 phone service'
 '24 phone service' '31' '54' '10 phone service' '11' '15' '32'
 '67 phone service' '29 phone service' '45' '16' '52' '21'
 '14 phone service' '68 phone service' '29' '69 phone service' '17' '46'
 '7' '66' '23 phone service' '37' '21 phone service' '27 phone service'
 '55' '42 phone service' '33 phone service' '69' '60' '13 phone service'

Tenure columns seems like it needs more cleaning

In [ ]:
#Applying a simple split method to extract numerical value
df2['tenure'] = df2['tenure'].apply(lambda a: a.split()[0])

In [ ]:
#checking values
df2['tenure'].unique()

array(['3', '53', '1', '56', '5', '48', '58', '10', '71', '65', '28',
       '67', '35', '72', '61', '70', '68', '47', '32', '13', '12', '9',
       '38', '42', '24', '27', '49', '50', '25', '22', '4', '18', '51',
       '62', '16', '8', '23', '31', '37', '30', '20', '36', '57', '41',
       '64', '26', '40', '34', '63', '44', '6', '33', '43', '59', '14',
       '19', '2', '66', '54', '11', '15', '29', '45', '52', '21', '69',
       '17', '46', '7', '55', '60', '39'], dtype=object)

In [ ]:
df2.head()

customerID  gender  ...  MonthlyCharges TotalCharges
0  9398-MMQTO    Male  ...           74.45       221.10
1  7619-PLRLP  Female  ...            74.1      3833.95
2  6457-GIRWB    Male  ...           69.35        69.35
4  1450-SKCVI  Female  ...           73.85      4092.85
5  4710-NKCAW    Male  ...            64.4       316.90

[5 rows x 9 columns]

In [ ]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1096 entries, 0 to 1367
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   customerID       1096 non-null   object 
 1   gender           1096 non-null   object 
 2   Senior Citizen   1096 non-null   int64  
 3   tenure           1096 non-null   object 
 4   InternetService  1096 non-null   object 
 5   Contract         1096 non-null   object 
 6   PaymentMethod    1096 non-null   object 
 7   MonthlyCharges   1096 non-null   object 
 8   TotalCharges     1096 non-null   float64
dtypes: float64(1), int64(1), object(7)
memory usage: 85.6+ KB


Tenure and monthly charges have to be changed to int64 and float format respectively

In [ ]:
df2['tenure'] = df2['tenure'].astype(int)

In [ ]:
df2['MonthlyCharges'] = df2['MonthlyCharges'].astype(float)

In [ ]:
#Just Checking
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1096 entries, 0 to 1367
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   customerID       1096 non-null   object 
 1   gender           1096 non-null   object 
 2   Senior Citizen   1096 non-null   int64  
 3   tenure           1096 non-null   int64  
 4   InternetService  1096 non-null   object 
 5   Contract         1096 non-null   object 
 6   PaymentMethod    1096 non-null   object 
 7   MonthlyCharges   1096 non-null   float64
 8   TotalCharges     1096 non-null   float64
dtypes: float64(2), int64(2), object(5)
memory usage: 85.6+ KB


### Mirroring manipulation to test data on training data

In [ ]:
df1.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

Removing columns that aren't in the test dataset

In [ ]:
df1.drop(['Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection','TechSupport', 'StreamingTV', 'StreamingMovies', 'PaperlessBilling'], axis = 1, inplace = True)

In [ ]:
df1.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'tenure', 'InternetService',
       'Contract', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [ ]:
df2.columns

Index(['customerID', 'gender', 'Senior Citizen', 'tenure', 'InternetService',
       'Contract', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges'],
      dtype='object')

In [ ]:
for i in df1.columns:
    for x in df2.columns:
        if i == x:
            print('True')

True
True
True
True
True
True
True
True


In [ ]:
df1.head()

customerID  gender  SeniorCitizen  ...  MonthlyCharges TotalCharges Churn
0  7590-VHVEG  Female              0  ...           29.85        29.85    No
1  5575-GNVDE    Male              0  ...           56.95       1889.5    No
2  3668-QPYBK    Male              0  ...           53.85       108.15   Yes
3  7795-CFOCW    Male              0  ...           42.30      1840.75    No
4  9237-HQITU  Female              0  ...           70.70       151.65   Yes

[5 rows x 10 columns]

In [ ]:
#Check for null values and remove those rows (display percentage though)
df1.isnull().sum() 

customerID         0
gender             0
SeniorCitizen      0
tenure             0
InternetService    0
Contract           0
PaymentMethod      0
MonthlyCharges     0
TotalCharges       0
Churn              0
dtype: int64

In [ ]:
for i in df2.columns:
  print(i, df2[i].unique())

customerID ['9398-MMQTO' '7619-PLRLP' '6457-GIRWB' ... '1941-HOSAM' '9110-HSGTV'
 '6171-ZTVYB']
gender ['Male' 'Female']
Senior Citizen [0 1]
tenure [ 3 53  1 56  5 48 58 10 71 65 28 67 35 72 61 70 68 47 32 13 12  9 38 42
 24 27 49 50 25 22  4 18 51 62 16  8 23 31 37 30 20 36 57 41 64 26 40 34
 63 44  6 33 43 59 14 19  2 66 54 11 15 29 45 52 21 69 17 46  7 55 60 39]
InternetService ['Fiber optic' 'DSL']
Contract ['Month-to-month' 'One year' 'Two year']
PaymentMethod ['Credit card (automatic)' 'Bank transfer (automatic)' 'Electronic check'
 'Mailed check']
MonthlyCharges [ 74.45  74.1   69.35  73.85  64.4   55.8   75.15  99.15  56.75  69.6
 104.15 110.8   80.15  35.75  69.9   89.2   55.65  50.7   30.5   98.3
  45.55 101.05 103.7   36.25  49.4  107.4   82.    45.05  64.55  86.25
  89.1   95.55  75.4  101.25 102.6   56.3   94.2   43.05  89.5   74.4
  74.35  99.75 111.95  94.    98.85  64.35  72.    49.7   80.7   24.2
  39.    65.45  83.2   25.    40.2   94.1  108.35  69.5   76.    93.6
  

In [ ]:
#Look at unique values and see if they need fixing (split fix)
for i in df1.columns:
  print(i, df1[i].unique())

customerID ['7590-VHVEG' '5575-GNVDE' '3668-QPYBK' ... '4801-JZAZL' '8361-LTMKD'
 '3186-AJIEK']
gender ['Female' 'Male']
SeniorCitizen [0 1]
tenure [ 1 34  2 45  8 22 10 28 62 13 16 58 49 25 69 52 71 21 12 30 47 72 17 27
  5 46 11 70 63 43 15 60 18 66  9  3 31 50 64 56  7 42 35 48 29 65 38 68
 32 55 37 36 41  6  4 33 67 23 57 61 14 20 53 26 54 24 44  0 40 39 59 51
 19]
InternetService ['DSL' 'Fiber optic' 'No']
Contract ['Month-to-month' 'One year' 'Two year']
PaymentMethod ['Electronic check' 'Mailed check' 'Bank transfer (automatic)'
 'Credit card (automatic)']
MonthlyCharges [29.85 56.95 53.85 ... 61.1  68.45 78.7 ]
TotalCharges ['29.85' '1889.5' '108.15' ... '346.45' '306.6' '6844.5']
Churn ['No' 'Yes']


In [ ]:
#Change data types if necessary
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5674 entries, 0 to 5673
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   customerID       5674 non-null   object 
 1   gender           5674 non-null   object 
 2   SeniorCitizen    5674 non-null   int64  
 3   tenure           5674 non-null   int64  
 4   InternetService  5674 non-null   object 
 5   Contract         5674 non-null   object 
 6   PaymentMethod    5674 non-null   object 
 7   MonthlyCharges   5674 non-null   float64
 8   TotalCharges     5674 non-null   object 
 9   Churn            5674 non-null   object 
dtypes: float64(1), int64(2), object(7)
memory usage: 443.4+ KB


In [ ]:
df1['TotalCharges'] = df1['TotalCharges'].astype(float)

ValueError: ignored

In [ ]:

for i in df1['TotalCharges']:
  try:
    float(i)
  except ValueError:
    print('value is', i)


value is  
value is  
value is  
value is  
value is  
value is  
value is  
value is  
value is  
value is  


In [ ]:
df1[df1['TotalCharges']==' ']

customerID  gender  SeniorCitizen  ...  MonthlyCharges TotalCharges Churn
236   4472-LVYGI  Female              0  ...           52.55                 No
501   3115-CZMZD    Male              0  ...           20.25                 No
684   5709-LVOEQ  Female              0  ...           80.85                 No
872   1371-DWPAZ  Female              0  ...           56.05                 No
2393  7644-OMVMY    Male              0  ...           19.85                 No
2779  3213-VVOLG    Male              0  ...           25.35                 No
3178  2520-SGTTA  Female              0  ...           20.00                 No
4016  2923-ARZLG    Male              0  ...           19.70                 No
5410  4075-WKNIU  Female              0  ...           73.35                 No
5494  2775-SEFEE    Male              0  ...           61.90                 No

[10 rows x 10 columns]

In [ ]:
#The error is being caused by a couple of spaces instead of values in the TotalCharges column
#I'm removing those rows
df1 = df1[df1['TotalCharges']!= ' ']

In [ ]:
#Now we try changing datatype
df1['TotalCharges'] = df1['TotalCharges'].astype(float)

In [ ]:
#Taking an example of where "No internet service" was repeated in a data row. Python was smart enough to rectify the error
df1[df1['customerID']=='2569-WGERO']

customerID  gender  SeniorCitizen  ...  MonthlyCharges TotalCharges Churn
5668  2569-WGERO  Female              0  ...           21.15       1419.4    No

[1 rows x 10 columns]